apps/src/panels/PanelsView.tsx

```diff
@@ -0,0 +1,163 @@
+// only used for levels that use Lab2.
+
+import React, {useCallback, useEffect, useLayoutEffect, useState} from 'react';
+import {useSelector} from 'react-redux';
+import {useAppDispatch} from '@cdo/apps/util/reduxHooks';
+import {
+  sendSuccessReport,
+  navigateToNextLevel,
+} from '@cdo/apps/code-studio/progressRedux';
+import {LabState} from '@cdo/apps/lab2/lab2Redux';
+import {PanelsLevelData} from './types';
+import styles from './panels.module.scss';
+import classNames from 'classnames';
+import EnhancedSafeMarkdown from '@cdo/apps/templates/EnhancedSafeMarkdown';
+import FontAwesome from '@cdo/apps/templates/FontAwesome';
+const commonI18n = require('@cdo/locale');
+
+function useWindowSize() {
+  const [size, setSize] = useState([
+    document.documentElement.clientWidth,
+    document.documentElement.clientHeight,
+  ]);
+  useLayoutEffect(() => {
+    function updateSize() {
+      const width = document.documentElement.clientWidth;
+      const height = document.documentElement.clientHeight;
+      setSize([width, height]);
+    }
+    window.addEventListener('resize', updateSize);
+    updateSize();
+    return () => window.removeEventListener('resize', updateSize);
+  }, []);
+  return size;
+}
+
+const PanelsView: React.FunctionComponent = () => {
+  const dispatch = useAppDispatch();
+
+  const appName = 'panels';
+
+  const levelData = useSelector(
+    ({lab}: {lab: LabState}) => lab.levelProperties?.levelData
+  );
+  const currentAppName = useSelector(
+    ({lab}: {lab: LabState}) => lab.levelProperties?.appName
+  );
+
+  const [levelPanels, setLevelPanels] = useState<PanelsLevelData | null>(null);
+  const [currentPanel, setCurrentPanel] = useState(0);
+
+  useEffect(() => {
+    if (currentAppName === appName && levelData) {
+      setLevelPanels(levelData as PanelsLevelData);
+    }
+  }, [currentAppName, levelData]);
+
+  const handleButtonClick = useCallback(() => {
+    if (levelPanels?.panels) {
+      if (currentPanel < levelPanels.panels.length - 1) {
+        setCurrentPanel(currentPanel + 1);
+      } else {
+        dispatch(sendSuccessReport(appName));
+        dispatch(navigateToNextLevel());
+      }
+    }
+  }, [levelPanels, currentPanel, dispatch]);
+
+  const handleBubbleClick = (index: number) => {
+    setCurrentPanel(index);
+  };
+
+  // Leave a margin to the left and the right of the panels, to the edges
+  // of the screen.
+  const horizontalMargin = 40;
+
+  // Leave a vertical margin above and below the panels, to the edges of the
+  // screen.
+  const verticalMargin = 50;
+
+  // We need room below the panels content for the children passed in.  This area
+  // can contain things like a Continue button.
+  const childrenAreaHeight = 70;
+
+  // The aspect ratio of the panels.
+  const contentAspectRatio = 16 / 9;
+
+  let [targetWidth, targetHeight] = useWindowSize();
+  targetWidth -= horizontalMargin * 2;
+  targetHeight -= verticalMargin * 2 + childrenAreaHeight;
+
+  let width, height;
```

Comment:
What do you think about memoizing the width/height calculation (using `useMemo`)? That way it only needs to be recomputed whenever `[targetWidth, targetHeight]` changes whereas now I believe it would recalculate on every render

Reply:
Good call; done.